File: src/Mapper.cpp
// Mapper.cpp
#include "../lib/mapper.h"

Mapper::Mapper() : prg_size(0), chr_size(0) {}

bool Mapper::cpuread(uint16_t address, uint16_t* map_address) {
    return false;
}

bool Mapper::cpuwrite(uint16_t address, uint16_t* map_address) {
    return false;
}

bool Mapper::ppuread(uint16_t address, uint16_t* map_address) {
    return false;
}

bool Mapper::ppuwrite(uint16_t address, uint16_t* map_address) {
    return false;
}

File: src/cartridge/mapper.rs
use super::MirrorMode;

pub trait Mapper{
    fn cpu_read(&self, address: u16,mapped_addr: &mut u32, data: &mut u8) -> bool;
    fn cpu_write(&mut self, address: u16,mapped_addr: &mut u32, data: u8) -> bool;
    fn ppu_read(&mut self, address: u16,mapped_addr: &mut u32, data: u8) -> bool;
    fn ppu_write(&mut self, address: u16,mapped_addr: &mut u32, data: u8) -> bool;
    fn get_mirror_mode(&self) -> MirrorMode;
    fn irq_clear(&mut self);
    fn savestate(&self);
    fn loadstate(&mut self);
    fn hasirq(&mut self) -> bool;
    fn scanline(&mut self);
    fn reset(&mut self);
}